contracts/build.go

```diff
@@ -0,0 +1,127 @@
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	DefaultBindPkg = "ethereum"
+)
+
+// ContractBuilder is an interface to build different chains contracts
+type ContractBuilder interface {
+	Targets() ([]*ContractTarget, error)
+	Compile(targets []*ContractTarget) error
+	GenerateBindings(targets []*ContractTarget) error
+}
+
+// ContractTarget contract build target
+type ContractTarget struct {
+	Version   string
+	Contracts []string
+}
+
+// abigenArgs is the arguments to the abigen executable. E.g., Bin is the -bin
+// arg.
+type abigenArgs struct {
+	ExecutablePath, Bin, ABI, Out, Type, Pkg string
+}
+
+// EthereumContractBuilder is a builder struct for Ethereum
+type EthereumContractBuilder struct {
+	ExecutablePath string
+}
+
+// NewEthereumContractBuilder builds solidity contracts
+func NewEthereumContractBuilder(execPath string) *EthereumContractBuilder {
+	return &EthereumContractBuilder{ExecutablePath: execPath}
+}
+
+// Compile compiles contracts
+func (b *EthereumContractBuilder) Compile(_ []*ContractTarget) error {
+	// step is not needed for Ethereum, already has ABI/BIN build in CI
+	return nil
+}
+
+// Targets what contracts we want to build
+func (b *EthereumContractBuilder) Targets() ([]*ContractTarget, error) {
+	return []*ContractTarget{
+		{
+			Contracts: []string{"LinkToken"},
+			Version:   "v0.4",
+		},
+		{
+			Contracts: []string{"APIConsumer", "BlockhashStore", "FluxAggregator",
+				"Oracle", "VRF", "VRFConsumer", "VRFCoordinator"},
+			Version: "v0.6",
+		},
+		{
+			Contracts: []string{"KeeperConsumer", "KeeperRegistry", "MockETHLINKAggregator",
+				"MockGASAggregator", "UpkeepRegistrationRequests", "OffchainAggregator"},
+			Version: "v0.7",
+		},
+	}, nil
+}
+
+// GenerateBindings generate go bindings for contracts
+func (b *EthereumContractBuilder) GenerateBindings(targets []*ContractTarget) error {
+	for _, t := range targets {
+		log.Debug().Str("Version", t.Version).Msg("Generating bindings for contracts")
+		for _, c := range t.Contracts {
+			err := abigen(abigenArgs{
+				ExecutablePath: b.ExecutablePath,
+				ABI:            filepath.Join(tools.EthereumContractsDir, t.Version, "abi", fmt.Sprintf("%s.abi", c)),
+				Bin:            filepath.Join(tools.EthereumContractsDir, t.Version, "bin", fmt.Sprintf("%s.bin", c)),
+				Out:            filepath.Join(tools.EthereumContractsDir, fmt.Sprintf("%s.go", c)),
+				Pkg:            c,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// commonBindPackage replaces individual pkg name with common name for convenience
+func commonBindPackage(filepath string, pkg string) error {
+	read, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		return err
+	}
+	newContents := strings.Replace(string(read), pkg, DefaultBindPkg, 1)
+	err = ioutil.WriteFile(filepath, []byte(newContents), 0)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// abigen calls abigen binary to generate bindings
+func abigen(a abigenArgs) error {
```

Comment:
Can use abigen programmatically, but need to keep in mind go-ethereum version in main repo.